1.0.0 release CHANGELOG update (#1481)
stellargraph/stellargraph: CHANGELOG.md

# Change Log

## [1.0.0](https://github.com/stellargraph/stellargraph/tree/v1.0.0)

[Full Changelog](https://github.com/stellargraph/stellargraph/compare/v0.11.0...v1.0.0)

> This 1.0 release of StellarGraph is the culmination of 2 years of activate development.

Reviewer: ```suggestion
This 1.0 release of StellarGraph is the culmination of 2 years of active development.
```

> This 1.0 release of StellarGraph is the culmination of 2 years of activate development.

Author: Could we use some of the comms info here to give a bigger announcement, eg:

```suggestion
This 1.0 release of StellarGraph is the culmination of three years of active research and engineering to deliver an open-source, user-friendly library for machine learning (ML) on graphs and networks.
```

  - Notebooks are [now published in the API documentation][demos-1.0.0], for better & faster rendering and more convenient access [\#1279](https://github.com/stellargraph/stellargraph/pull/1279) [\#1433](https://github.com/stellargraph/stellargraph/pull/1433) [\#1448](https://github.com/stellargraph/stellargraph/pull/1448)
  - The [demos indices][demos-1.0.0] and [READMEs](demos/) now contain more guidance and explanation to make it easier to find a relevant example [\#1200](https://github.com/stellargraph/stellargraph/pull/1200)
  - Several demos have been added or rewritten: [loading data from Neo4j][neo4j-1.0.0] [\#1184](https://github.com/stellargraph/stellargraph/pull/1184), [link prediction using Node2Vec][n2v-lp-1.0.0] [\#1190](https://github.com/stellargraph/stellargraph/pull/1190), [graph classification with GCN][gc-gcn-1.0.0], [graph classification with DGCNN][gc-dgcnn-1.0.0]
  - Notebooks now detect if they're being used with an incorrect version of the StellarGraph library, eliminating confusion about version mismatches [\#1242](https://github.com/stellargraph/stellargraph/pull/1242)

>   - Notebooks are easier to download, both individually via a button on each in the API documentation [\#1460](https://github.com/stellargraph/stellargraph/pull/1460) and in bulk via [a `curl` command]() [\#1377](https://github.com/stellargraph/stellargraph/pull/1377) [\#1459](https://github.com/stellargraph/stellargraph/pull/1459)

Reviewer: Is the `curl` command supposed to be pointing to a link? Otherwise can be
```suggestion
  - Notebooks are easier to download, both individually via a button on each in the API documentation [\#1460](https://github.com/stellargraph/stellargraph/pull/1460) and in bulk via a `curl` command [\#1377](https://github.com/stellargraph/stellargraph/pull/1377) [\#1459](https://github.com/stellargraph/stellargraph/pull/1459)
```

Reviewer 2: Oh, yeah, I forgot. I was going to link to https://stellargraph.readthedocs.io/en/v1.0.0/demos/index.html#download-the-demos (i.e. the 1.0.0 version of `develop`'s https://stellargraph.readthedocs.io/en/latest/demos/index.html#download-the-demos).

This reminds me: should all of our other links to demos be to https://stellargraph.readthedocs.io/en/v1.0.0/demos/... rather than the relative links?

Reviewer: > This reminds me: should all of our other links to demos be to https://stellargraph.readthedocs.io/en/v1.0.0/demos/... rather than the relative links?

Yeah that's probably better, I think the experience has become a lot nicer on rtd so should just try to point there where possible. I'll update the links

>   - Notebooks are easier to download, both individually via a button on each in the API documentation [\#1460](https://github.com/stellargraph/stellargraph/pull/1460) and in bulk via [a `curl` command]() [\#1377](https://github.com/stellargraph/stellargraph/pull/1377) [\#1459](https://github.com/stellargraph/stellargraph/pull/1459)

Author: I think curl here is too much detail?
```suggestion
  - Notebooks are easier to download, both individually via a button on each in the API documentation [\#1460](https://github.com/stellargraph/stellargraph/pull/1460) and in bulk [\#1377](https://github.com/stellargraph/stellargraph/pull/1377) [\#1459](https://github.com/stellargraph/stellargraph/pull/1459)
```

Reviewer: Yeah I'll just remove it 👍

  - any model using `FullBatchNodeGenerator`, including models supported in StellarGraph 0.11: `GCN`, `GAT`, `PPNP` and `APPNP`
  - `GraphSAGE` [\#1162](https://github.com/stellargraph/stellargraph/pull/1162)
  - `HinSAGE` for heterogeneous graphs with node features [\#1254](https://github.com/stellargraph/stellargraph/pull/1254)
- `UnsupervisedSampler` supports a `walker` parameter to use other random walking algorithms such as `BiasedRandomWalk`, in addition to the default `UniformRandomWalk`. [\#1187](https://github.com/stellargraph/stellargraph/pull/1187)
- The `StellarGraph` class is now smaller, faster and easier to construct and use:
  - The `StellarGraph(..., edge_type_column=...)` parameter can be used to construct a heterogeneous graph from a single flat `DataFrame`, containing a column of the edge types [\#1284](https://github.com/stellargraph/stellargraph/pull/1284). This avoids the need to build separate `DataFrame`s for each type, and is significantly faster when there are many types. Using `edge_type_column` gives a 2.6× speedup for loading the `stellargraph.datasets.FB15k` dataset (with almost 600 thousand edges across 1345 types).

>   - `StellarGraph`'s internal cache of node adjacencies is now computed lazily [\#1291](https://github.com/stellargraph/stellargraph/pull/1291) and takes into account whether the graph is directed or not [\#1463](https://github.com/stellargraph/stellargraph/pull/1463) (up to 63% smaller, 30× speedup on a reddit dataset with 11.6 million edges), and they now uses the smallest integer type it can [\#1289](https://github.com/stellargraph/stellargraph/pull/1289) (36% smaller on the reddit dataset)

Reviewer: ```suggestion
  - `StellarGraph`'s internal cache of node adjacencies is now computed lazily [\#1291](https://github.com/stellargraph/stellargraph/pull/1291) and takes into account whether the graph is directed or not [\#1463](https://github.com/stellargraph/stellargraph/pull/1463) (up to 63% smaller, 30× speedup on a reddit dataset with 11.6 million edges), and they now use the smallest integer type they can [\#1289](https://github.com/stellargraph/stellargraph/pull/1289) (36% smaller on the reddit dataset)
```

  - `StellarGraph`'s internal list of source and target nodes are now stored using integer "ilocs" [\#1267](https://github.com/stellargraph/stellargraph/pull/1267), reducing memory use and making some functionality significantly faster (selection of speedups: `StellarGraph.to_adjacency_matrix` 1.6×, GraphSAGE sampling 1.4×, weighted `BiasedRandomWalk` 2.4× [\#1444](https://github.com/stellargraph/stellargraph/pull/1444), `UniformRandomMetaPathWalk` 6× [\#1446](https://github.com/stellargraph/stellargraph/pull/1446))
  - Functions like `graph.node_features()` no longer needs `node_type` specified if `graph` has only one node type (this includes classes like `HinSAGENodeGenerator`, which no longer needs `head_node_type` if there is only one node type) [\#1375](https://github.com/stellargraph/stellargraph/pull/1375)

[demos-1.0.0]: https://stellargraph.readthedocs.io/en/v1.0.0/demos/index.html

### Breaking changes

- The `stellargraph/stellargraph` docker image wasn't being published in an optimal way, so we have stopped updating it for now [\#1455](https://github.com/stellargraph/stellargraph/pull/1455)

> - Edge weights are now validated to be numeric when creating a `StellarGraph`, previously edge weights could be any type, but all algorithms that use them would fail. [\#1191](https://github.com/stellargraph/stellargraph/pull/1191)

Reviewer: ```suggestion
- Edge weights are now validated to be numeric when creating a `StellarGraph`. Previously edge weights could be any type, but all algorithms that use them would fail with non-numeric types. [\#1191](https://github.com/stellargraph/stellargraph/pull/1191)
```

# Change Log

## [1.0.0](https://github.com/stellargraph/stellargraph/tree/v1.0.0)

[Full Changelog](https://github.com/stellargraph/stellargraph/compare/v0.11.0...v1.0.0)

This 1.0 release of StellarGraph is the culmination of three years of active research and engineering to deliver an open-source, user-friendly library for machine learning (ML) on graphs and networks.

Jump in to this release, with the new demos and examples:

> - [More helpful indexing and guidance in demo READMEs](demos/)

Reviewer: Or just switch to the readthedocs link?
```suggestion
- [More helpful indexing and guidance for demos in our API documentation](https://stellargraph.readthedocs.io/en/latest/demos/index.html)
```

# Change Log

## [1.0.0](https://github.com/stellargraph/stellargraph/tree/v1.0.0)

[Full Changelog](https://github.com/stellargraph/stellargraph/compare/v0.11.0...v1.0.0)

This 1.0 release of StellarGraph is the culmination of three years of active research and engineering to deliver an open-source, user-friendly library for machine learning (ML) on graphs and networks.

Jump in to this release, with the new demos and examples:

- [More helpful indexing and guidance in demo READMEs](demos/)
- [Loading from Neo4j][neo4j-1.0.0]
- [More explanatory Node2Vec link prediction][n2v-lp-1.0.0]
- [Unsupervised `GraphSAGE` and `HinSAGE` via `DeepGraphInfomax`][dgi-1.0.0]
- Graph classification [with `GCNSupervisedGraphClassification`][gc-gcn-1.0.0] and [with `DeepGraphCNN`][gc-dgcnn-1.0.0]
- [Time series prediction using spatial information, using `GraphConvolutionLSTM`][gcn-lstm-1.0.0] (experimental)

> [neo4j-1.0.0]: demos/basics/loading-saving-neo4j.ipynb

Author: I think all of these notebooks will need to link to readthedocs?

Reviewer: Yep I'm going to switch the links now that that one has landed

### Major features and improvements

- Better demonstration notebooks and documentation to make the library more accessible to new and existing users:

>   - Notebooks are [now published in the API documentation][demos-1.0.0], for better & faster rendering and more convenient access [\#1279](https://github.com/stellargraph/stellargraph/pull/1279) [\#1433](https://github.com/stellargraph/stellargraph/pull/1433) [\#1448](https://github.com/stellargraph/stellargraph/pull/1448). Note that some notebooks in the repo are currently omitted from the API docs until they're reviewed further [\#1477](https://github.com/stellargraph/stellargraph/pull/1477)

Reviewer 2: I think it's probably ok to not mention the non-visible notebooks here? It seems a bit like the wrong foot start with.

```suggestion
  - Notebooks are [now published in the API documentation][demos-1.0.0], for better & faster rendering and more convenient access [\#1279](https://github.com/stellargraph/stellargraph/pull/1279) [\#1433](https://github.com/stellargraph/stellargraph/pull/1433) [\#1448](https://github.com/stellargraph/stellargraph/pull/1448).
```

Or maybe:

```suggestion
  - Most notebooks are [now published in the API documentation][demos-1.0.0], for better & faster rendering and more convenient access [\#1279](https://github.com/stellargraph/stellargraph/pull/1279) [\#1433](https://github.com/stellargraph/stellargraph/pull/1433) [\#1448](https://github.com/stellargraph/stellargraph/pull/1448).
```

Reviewer: Yeah good point, I'll just remove it

  - `StellarGraph`'s internal list of source and target nodes are now stored using integer "ilocs" [\#1267](https://github.com/stellargraph/stellargraph/pull/1267), reducing memory use and making some functionality significantly faster (selection of speedups: `StellarGraph.to_adjacency_matrix` 1.6×, GraphSAGE sampling 1.4×, weighted `BiasedRandomWalk` 2.4× [\#1444](https://github.com/stellargraph/stellargraph/pull/1444), `UniformRandomMetaPathWalk` 6× [\#1446](https://github.com/stellargraph/stellargraph/pull/1446))
  - Functions like `graph.node_features()` no longer needs `node_type` specified if `graph` has only one node type (this includes classes like `HinSAGENodeGenerator`, which no longer needs `head_node_type` if there is only one node type) [\#1375](https://github.com/stellargraph/stellargraph/pull/1375)

> - Performance and memory usage improvements in numbers:

Reviewer 2: To make this a little clearer what the comparison is:

```suggestion
- Overall performance and memory usage improvements since 0.11, in numbers:
```

> - Performance and memory usage improvements in numbers:

Reviewer 2: Maybe all of the performance numbers here could be collated slightly, e.g. remove the two reddit dataset and "(selection of speedups: ...)" paratheticals above and put all of the performance numbers here. A possibility might be:

```markdown
- Performance and memory usage improvements in numbers: (or the other title)
  - The FB15k graph has 15 thousand nodes and 483 thousand edges: it is now 7× faster and 4× smaller to construct (without adjacency lists). It is still about 2× smaller when directed or undirected adjacency lists are computed.
  - Directed adjacency matrix construction is 1.6× faster
  - Various samplers and random walkers are faster: `HinSAGENodeGenerator` is 3× faster (on `MovieLens`), `Attri2VecNodeGenerator` is 4× faster (on `CiteSeer`), weighted `BiasedRandomWalk` is up to 3× faster, `UniformRandomMetapathWalk` is up to 7× faster
```

Rendered view:

> - Performance and memory usage improvements in numbers: (or the other title)
>   - The FB15k graph has 15k nodes and 483k edges: it is now 7× faster and 4× smaller to construct (without adjacency lists). It is also about 2× smaller after directed or undirected adjacency lists are computed.
>   - Directed adjacency matrix construction is 1.6× faster
>   - Various samplers and random walkers are faster: `HinSAGENodeGenerator` is 3× faster (on `MovieLens`), `Attri2VecNodeGenerator` is 4× faster (on `CiteSeer`), weighted `BiasedRandomWalk` is up to 3× faster, `UniformRandomMetapathWalk` is up to 7× faster

>   - Base `StellarGraph` is up to 4.2 x smaller (tested on FB15K)

Reviewer 2: Minor but `×` (https://en.wikipedia.org/wiki/Multiplication_sign#Unicode_and_HTML_entities) looks slightly better than x

```suggestion
  - Base `StellarGraph` is up to 4.2× smaller (tested on FB15K)
```

similarly for the rest of these `x`s.

  - Base `StellarGraph` construction is 7 x faster (tested on FB15K)
  - Undirected `StellarGraph` is up to 2.2 x smaller (tested on FB15K)
  - Directed `StellarGraph` is up to 1.8 x smaller (tested on FB15K)
  - Directed adjacency matrix construction is 1.6 x faster (tested on synthetic benchmark)
  - `HinSAGENodeGenerator` is 2.9 x faster (tested on MovieLens)
  - `Attri2VecNodeGenerator` is 4.3 x faster (tested on CiteSeer)
  - Weighted `BiasedRandomWalk` is 2.6 x faster (tested on synthetic benchmark)
  - `UniformRandomMetapathWalk` is 6.6 x faster (tested on synthetic benchmark)

[demos-1.0.0]: https://stellargraph.readthedocs.io/en/v1.0.0/demos/index.html

> ### Breaking changes

Reviewer 2: Oh, one more potential breaking change: https://github.com/stellargraph/stellargraph/pull/1444 goes from collapsing all multi-edges to considering each separately (summing the weights).

Specifically: it previously required all multi edges had the same weight, and only counted one of them when considering where to walk, so a graph like `a --[1]-- b` (twice times), `a --[1]--- c` would be equivalent to the only having the first edge once, and thus a random walk from `a` would have equal chance of going to `b` or `c` (without biases). Now it is equivalent to having an `a --[2]-- b`, and so `b` is twice as likely as `c`.

Worth mentioning?

Reviewer: Hmm right, would something like below be clear enough?
```
- `BiasedRandomWalk` now takes multi-edges into consideration instead of collapsing them when traversing the graph.
```
Or could spell it out a bit more:
```
- `BiasedRandomWalk` now takes multi-edges into consideration instead of collapsing them when traversing the graph. It previously required all multi-edges had to same weight and only counted one of them when considering where to walk, but now a multi-edge is equivalent to having an edge whose weight is the sum of the weights of all edges in the multi-edge.
```

Reviewer 2: The first one is good; could potentially expand very slightly

> - `BiasedRandomWalk` now takes multi-edges into consideration (by effectively summing weights) instead of collapsing them when traversing the graph.

Reviewer: (I've put in the long form)

Reviewer 2: That works for me 👍

- Some functions no longer fail for some particular cases of empty graphs: `StellarGraph.to_adjacency_matrix` [\#1378](https://github.com/stellargraph/stellargraph/pull/1378), `StellarGraph.from_networkx` [\#1401](https://github.com/stellargraph/stellargraph/pull/1401)
- `CorruptedGenerator` on a `FullBatchNodeGenerator` can be used to train `DeepGraphInfomax` on a subset of the nodes in a graph, instead of all of them [\#1415](https://github.com/stellargraph/stellargraph/pull/1415)
- The `stellargraph.custom_keras_layers` dictionary for use when loading a Keras model now includes all of StellarGraph's layers [\#1280](https://github.com/stellargraph/stellargraph/pull/1280)
- `PaddedGraphGenerator.flow` now also accepts a list of `StellarGraph` objects as input [\#1458](https://github.com/stellargraph/stellargraph/pull/1458)
- Supervised Graph Classification demo now prints progress update messages during training [\#1485](https://github.com/stellargraph/stellargraph/pull/1485)

> - Explicit contributors file has been removed in order to avoid potentially having inconsistent acknowledgement [\#1484](https://github.com/stellargraph/stellargraph/pull/1484). Please refer to the [Github display for contributors](https://github.com/stellargraph/stellargraph/graphs/contributors) instead.

Reviewer 2: 👍 good idea

We did have inconsistent acknowledgement, so:

```suggestion
- Explicit contributors file has been removed to avoid inconsistent acknowledgement [\#1484](https://github.com/stellargraph/stellargraph/pull/1484). Please refer to the [Github display for contributors](https://github.com/stellargraph/stellargraph/graphs/contributors) instead.
```
